include/nbl/asset/ICPUDescriptorSet.h

```diff
@@ -25,295 +25,105 @@ namespace nbl::asset
 		}
+	}
 
-        core::smart_refctd_ptr<IAsset> clone(uint32_t _depth = ~0u) const override
-        {
-            auto layout = (_depth > 0u && m_layout) ? core::smart_refctd_ptr_static_cast<ICPUDescriptorSetLayout>(m_layout->clone(_depth - 1u)) : m_layout;
-            auto cp = core::make_smart_refctd_ptr<ICPUDescriptorSet>(std::move(layout));
-            clone_common(cp.get());
+	_NBL_STATIC_INLINE_CONSTEXPR auto AssetType = ET_DESCRIPTOR_SET;
+	inline E_TYPE getAssetType() const override { return AssetType; }
 
-			auto cloneDescriptor = [](const core::smart_refctd_ptr<IDescriptor>& _desc, uint32_t _depth) -> core::smart_refctd_ptr<IDescriptor>
-			{
-				if (!_desc)
-					return nullptr;
+	inline ICPUDescriptorSetLayout* getLayout() 
+	{
+		assert(!isImmutable_debug());
+		return m_layout.get();
+	}
 
-				IAsset* asset = nullptr;
-				switch (_desc->getTypeCategory())
-				{
-				case IDescriptor::EC_BUFFER:
-					asset = static_cast<ICPUBuffer*>(_desc.get()); break;
-				case IDescriptor::EC_BUFFER_VIEW:
-					asset = static_cast<ICPUBufferView*>(_desc.get()); break;
-				case IDescriptor::EC_IMAGE:
-					asset = static_cast<ICPUImageView*>(_desc.get()); break;
-				}
+	inline const ICPUDescriptorSetLayout* getLayout() const { return m_layout.get(); }
 
-				auto cp = asset->clone(_depth);
+	inline bool canBeRestoredFrom(const IAsset* _other) const override
+	{
+		auto* other = static_cast<const ICPUDescriptorSet*>(_other);
+		return m_layout->canBeRestoredFrom(other->m_layout.get());
+	}
 
-				switch (_desc->getTypeCategory())
-				{
-				case IDescriptor::EC_BUFFER:
-					return core::smart_refctd_ptr_static_cast<ICPUBuffer>(std::move(cp));
-				case IDescriptor::EC_BUFFER_VIEW:
-					return core::smart_refctd_ptr_static_cast<ICPUBufferView>(std::move(cp));
-				case IDescriptor::EC_IMAGE:
-					return core::smart_refctd_ptr_static_cast<ICPUImageView>(std::move(cp));
-				}
-				return nullptr;
-			};
+	inline size_t conservativeSizeEstimate() const override
+	{
+		assert(!"Invalid code path.");
+		return 0xdeadbeefull;
+	}
 
-			for (uint32_t t = 0u; t < EDT_COUNT; ++t)
-			{
-				const auto type = static_cast<E_DESCRIPTOR_TYPE>(t);
+	inline core::SRange<SDescriptorInfo> getDescriptorInfoStorage(const IDescriptor::E_TYPE type) const
+	{
+		if (!m_descriptorInfos[static_cast<uint32_t>(type)])
+			return { nullptr, nullptr };
+		else
+			return { m_descriptorInfos[static_cast<uint32_t>(type)]->begin(), m_descriptorInfos[static_cast<uint32_t>(type)]->end() };
+	}
 
-				for (uint32_t i = 0u; i < m_descriptors[type]->size(); ++i)
-				{
-					const auto& srcDescriptor = getDescriptorStorage(type)[i];
-					const auto& srcDescriptorInfo = getDescriptorInfoStorage(type)[i];
+	core::SRange<SDescriptorInfo> getDescriptorInfos(const ICPUDescriptorSetLayout::CBindingRedirect::binding_number_t binding, IDescriptor::E_TYPE type = IDescriptor::E_TYPE::ET_COUNT);
```

Comment:
again, you need a `const` and non-`const` version for mutability reasons

Reply:
Yep, can be done here.

Resolved.